hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractContractBulkDeleteTest.java

```diff
@@ -0,0 +1,255 @@
+                .describedAs("Path capability BULK_DELETE should be declared")
+                .isTrue();
+    }
+
+    @Test
+    public void testDeletePathsNotUnderBase() throws Exception {
+        List<Path> paths = new ArrayList<>();
+        Path pathNotUnderBase = path("not-under-base");
+        paths.add(pathNotUnderBase);
+        // Should fail as path is not under the base path.
+        intercept(IllegalArgumentException.class,
+                () -> bulkDelete(getFileSystem(), basePath, paths));
+    }
+
+    @Test
+    public void testDeletePathsNotAbsolute() throws Exception {
+        List<Path> paths = new ArrayList<>();
+        Path pathNotAbsolute = new Path("not-absolute");
+        paths.add(pathNotAbsolute);
+        // Should fail as path is not absolute.
+        intercept(IllegalArgumentException.class,
+                () -> bulkDelete(getFileSystem(), basePath, paths));
+    }
+
+    @Test
+    public void testDeletePathsNotExists() throws Exception {
+        List<Path> paths = new ArrayList<>();
+        Path pathNotExists = new Path(basePath, "not-exists");
+        paths.add(pathNotExists);
+        // bulk delete call doesn't verify if a path exist or not before deleting.
+        assertSuccessfulBulkDelete(bulkDelete(getFileSystem(), basePath, paths));
+    }
+
+    @Test
+    public void testDeletePathsDirectory() throws Exception {
+        List<Path> paths = new ArrayList<>();
+        Path dirPath = new Path(basePath, "dir");
+        fs.mkdirs(dirPath);
+        paths.add(dirPath);
+        Path filePath = new Path(dirPath, "file");
+        touch(fs, filePath);
+        paths.add(filePath);
+        pageSizePreconditionForTest(paths.size());
+        // Outcome is undefined. But call shouldn't fail. In case of S3 directories will still be present.
+        assertSuccessfulBulkDelete(bulkDelete(getFileSystem(), basePath, paths));
+    }
+
+    @Test
+    public void testDeleteEmptyDirectory() throws Exception {
+        List<Path> paths = new ArrayList<>();
+        Path emptyDirPath = new Path(basePath, "empty-dir");
+        fs.mkdirs(emptyDirPath);
+        paths.add(emptyDirPath);
+        // Should pass as empty directory.
+        assertSuccessfulBulkDelete(bulkDelete(getFileSystem(), basePath, paths));
+    }
+
+    @Test
+    public void testDeleteEmptyList() throws Exception {
+        List<Path> paths = new ArrayList<>();
+        // Empty list should pass.
+        assertSuccessfulBulkDelete(bulkDelete(getFileSystem(), basePath, paths));
+    }
+
+    @Test
+    public void testDeleteSamePathsMoreThanOnce() throws Exception {
+        List<Path> paths = new ArrayList<>();
+        Path path = new Path(basePath, "file");
+        touch(fs, path);
+        paths.add(path);
+        paths.add(path);
+        Path another = new Path(basePath, "another-file");
+        touch(fs, another);
+        paths.add(another);
+        pageSizePreconditionForTest(paths.size());
```

Comment:
build the paths and this precondition before the touch; saves any network IO before skipping